File: scripts/t-wrapper/translation-applier.ts
import { NodePath } from "@babel/traverse";
import * as t from "@babel/types";
import { ScriptConfig } from "../common/default-config";
import { generateCode } from "../common/ast/parser-utils";
import {
  hasTranslationFunctionCall,
  createTranslationBinding,
} from "./ast-helpers";
import { ensureNamedImport, ensureUseClientDirective } from "./import-manager";
import { STRING_CONSTANTS } from "./constants";
import { writeFile } from "./fs-utils";

/**
 * AST에 번역 바인딩 및 import 추가
 * 파일 쓰기는 포함하지 않음 (테스트 용이성을 위해)
 */
export function applyTranslationsToAST(
  ast: t.File,
  modifiedComponentPaths: NodePath<t.Function>[],
  config: Required<ScriptConfig>
): void {
  const isServerMode = config.mode === "server";
  const isClientMode = config.mode === "client";
  const isNextjsFramework = config.framework === "nextjs";

  if (isNextjsFramework && isClientMode) {
    ensureUseClientDirective(ast);
  }

  const usedTranslationFunctions = new Set<string>();

  // 수정된것 중 t가 있는 것을 찾아서 수정
  modifiedComponentPaths.forEach((componentPath) => {
    if (componentPath.scope.hasBinding(STRING_CONSTANTS.TRANSLATION_FUNCTION)) {
      return;
    }

    const body = componentPath.get("body");

    const translationFunctionName = isServerMode
      ? config.serverTranslationFunction
      : STRING_CONSTANTS.USE_TRANSLATION;

    if (hasTranslationFunctionCall(body, translationFunctionName)) {
      return;
    }

    if (isServerMode) {
      (componentPath.node as any).async = true;
    }
    const decl = createTranslationBinding(
      isServerMode ? "server" : "client",
      isServerMode ? config.serverTranslationFunction : undefined
    );

    // body 최상단에 추가
    if (body.isBlockStatement()) {
      body.unshiftContainer("body", decl);
    }
    // body 최상단이 아닌 경우 리턴 앞에 추가
    // 예시: const Home = () => <div />;
    else {
      const original = body.node as t.Expression;
      (componentPath.node as any).body = t.blockStatement([
        decl,
        t.returnStatement(original),
      ]);
    }

    // 사용된 번역 함수 추가
    usedTranslationFunctions.add(translationFunctionName);
  });

  // 사용된 번역 함수 가져와서 임포트 구문 추가
  usedTranslationFunctions.forEach((functionName) => {
    ensureNamedImport(ast, config.translationImportSource, functionName);
  });
}

/**
 * AST를 코드로 변환하여 파일에 쓰기
 */
export function writeASTToFile(
  ast: t.File,
  filePath: string,
  config: Required<ScriptConfig>
): void {
  const output = generateCode(ast, config.parserType, {
    retainLines: true,
    comments: true,
  });

  writeFile(filePath, output.code);
}

File: scripts/t-wrapper-rust/src/applier/translation_applier.rs
/// 번역 적용 로직
/// TypeScript의 translation-applier.ts와 동일

use swc_ecma_ast::*;
use anyhow::Result;
use std::fs;
use crate::parser::generate_code;

/// 설정 구조체 (간단 버전)
#[derive(Debug, Clone)]
pub struct ApplierConfig {
    pub mode: Option<String>,
    pub framework: Option<String>,
    pub server_translation_function: Option<String>,
    pub translation_import_source: String,
}

/// AST에 번역 바인딩 및 import 추가
/// 
/// TypeScript 버전과 동일한 로직:
/// 1. Next.js + client 모드면 'use client' 디렉티브 추가
/// 2. 수정된 컴포넌트들에 t 바인딩 추가
/// 3. 필요한 import 추가
pub fn apply_translations_to_ast(
    _ast: &mut Module,
    _modified_functions: &[String],
    _config: &ApplierConfig,
) {
    // TODO: SWC AST로 구현
    // TypeScript 버전의 로직:
    // 1. isNextjsFramework && isClientMode이면 ensureUseClientDirective
    // 2. modifiedComponentPaths.forEach로 각 컴포넌트 처리
    //    - scope.hasBinding("t") 확인
    //    - hasTranslationFunctionCall 확인
    //    - isServerMode면 async true + createTranslationBinding("server")
    //    - isClientMode면 createTranslationBinding("client")
    //    - body.unshiftContainer 또는 blockStatement로 감싸기
    // 3. usedTranslationFunctions.forEach로 import 추가
}

/// AST를 코드로 변환하여 파일에 쓰기
/// 
/// TypeScript 버전과 동일한 로직:
/// 1. generateCode로 AST를 코드로 변환
/// 2. fs::write로 파일에 쓰기
pub fn write_ast_to_file(
    ast: &Module,
    file_path: &str,
) -> Result<()> {
    let output = generate_code(ast)?;
    fs::write(file_path, output)?;
    Ok(())
}

